crates/cargo-test-macro/src/lib.rs

```diff
@@ -144,18 +147,61 @@ pub fn cargo_test(attr: TokenStream, item: TokenStream) -> TokenStream {
                     continue;
                 }
             }
+            TokenTree::Ident(i) => {
+                // The first time through it will be `fn` the second time is the
+                // name of the test.
+                if test_name.is_none() && num == 1 {
+                    test_name = Some(i.to_string())
+                } else {
+                    num += 1;
+                }
+                ret.extend(Some(TokenTree::Ident(i)));
+                continue;
+            }
             other => {
                 ret.extend(Some(other));
                 continue;
             }
         };
 
-        let mut new_body = to_token_stream(
-            r#"let _test_guard = {
+        let name = &test_name
+            .clone()
+            .map(|n| n.split("::").next().unwrap().to_string())
+            .unwrap();
+        let mut new_body = to_token_stream(&format!(
+            r#"let _test_guard = {{
                 let tmp_dir = option_env!("CARGO_TARGET_TMPDIR");
-                cargo_test_support::paths::init_root(tmp_dir)
-            };"#,
-        );
+                // looks at `file!()` to figure out where the test's directory should be.
+                // test fn `workspaces::workspace_in_git`
+                //   -> `file!` expands to `tests/testsuite/workspaces.rs`
+                //   -> `testsuite/workspaces`
+                // test fn `cargo_remove::cargo_remove::invalid_arg::case`
+                //   -> `file!` expands to `tests/testsuite/cargo_remove/invalid_arg.rs`
+                //   -> `testsuite/cargo_remove/invalid_arg`
+                // test fn `build-std::cross_custom`
+                //   -> `file!` expands to `tests/build-std/main.rs`
+                //   -> `build-std/main`
+                let test_dir: std::path::PathBuf = std::path::PathBuf::from(std::file!())
+                    .components()
+                    // skips `tests`
+                    .skip(1)
```

Comment:
Should we instead search forward/backwards for `tests`?  The location of `tests` depends on this being run within a workspace or not (and where in the workspace it is).